### DISCUSSION TITLE
failed to unmarshal yaml using composite workflows

### DISCUSSION THREAD

**Discussion by askb** [community/community#65057] [Actions] [answered]
### Github actions

failed to unmarshal yaml using composite workflows, ways to debug the workflows?

### Body

I am trying to get a job that is working well on a local repo to work with composite reusable workflows using actions. I am running the job (packer-validator) as a part of the workflow [1.] and also trying to get this working with reusable composite workflow [2.]. The job in question uses the composite workflow returns `failed to unmarshal yaml` errors [4.]

```
 ++ packer validate -var-file=/home/runner/work/releng-builder/releng-builder/cloud-env.pkrvars.hcl -var-file=common-packer/vars/centos-7-arm64.pkrvars.hcl templates/builder.pkr.hcl
  + output='Error: 1 error(s) occurred:
  
  * unable to load clouds.yaml: failed to unmarshal yaml: yaml: unmarshal errors:
    line 1: cannot unmarshal !!str `cloud_a...` into clientconfig.Clouds
  
    on templates/builder.pkr.hcl line 115:
  templates/builder.pkr.hcl: Error: 1 error(s) occurred:
    (source code not available)'
  + echo 'templates/builder.pkr.hcl: Error: 1 error(s) occurred:
  
  * unable to load clouds.yaml: failed to unmarshal yaml: yaml: unmarshal errors:
    line 1: cannot unmarshal !!str `cloud_a...` into clientconfig.Clouds
  
    on templates/builder.pkr.hcl line 115:
  
  * unable to load clouds.yaml: failed to unmarshal yaml: yaml: unmarshal errors:
    line 1: cannot unmarshal !!str `cloud_a...` into clientconfig.Clouds
  
    (source code not available)'
    on templates/builder.pkr.hcl line 115:
    (source code not available)
  + exit 1
```

IMU the only difference between the jobs is the secret that is being passed from the downstream repository that uses composite workflows [2.] to the upstream repo (diff organizations) might be causing issues. 

Upstream composite workflow is available here:
https://github.com/lfit/releng-reusable-workflows/blob/main/.github/workflows/compose-packer-verify.yaml

Appreciate any suggestions or insights on how to debug this!

[1.] https://github.com/opendaylight/releng-builder/actions/runs/5981012157/workflow#L147-L253
[2.] https://github.com/opendaylight/releng-builder/actions/runs/5981012172/workflow#L52-L70
[3.] https://github.com/opendaylight/releng-builder/actions/runs/5981012157/job/16228177321
[4.] https://github.com/opendaylight/releng-builder/actions/runs/5981012172/job/16228176559

**Comment by onurtufekci34** [comment]
The error message you're encountering indicates that there is a problem with unmarshaling a YAML configuration in your job, specifically related to a `clouds.yaml` file. To debug this issue in your GitHub Actions workflow, you can follow these steps:

1. **Check YAML Syntax**: Ensure that the YAML syntax in your composite workflow and job configuration files is correct. YAML is sensitive to indentation and formatting, so any incorrect structure can lead to parsing errors.

2. **Review YAML File**: Examine the content of the `clouds.yaml` file in your repository. Ensure it is well-formed and free of any syntax errors or unexpected values.

3. **Local Testing**: Verify that your job (packer-validator) works well on your local machine. Make sure it can successfully access and load the `clouds.yaml` file. This helps ensure that the issue is specific to the GitHub Actions environment.

4. **Environment Variables**: Confirm that any environment variables used in your job are correctly set in your GitHub Actions workflow. You can echo these variables in your workflow to check their values.

   ```yaml
   - name: Debug Environment Variables
     run: echo "Clouds YAML: $CLOUDS_YAML"
   ```

5. **Debug Logging**: Add additional logging and debug statements in your workflow to get more information about the environment and variables at runtime. For example, you can print the contents of the `clouds.yaml` file and other relevant variables.

   ```yaml
   - name: Debug Clouds YAML
     run: |
       cat /path/to/clouds.yaml
       echo "Variable Value: $VARIABLE_NAME"
   ```

6. **Secrets and Sensitive Data**: If your `clouds.yaml` file contains sensitive information, make sure that you're handling secrets properly using GitHub Secrets. Avoid printing or exposing sensitive data in the logs.

7. **Artifact Workflow**: If your composite workflow relies on artifacts from other workflows, make sure that the artifacts are being passed correctly and are available for the job.

8. **Use Checkout Action**: Ensure you've used the `actions/checkout` action to checkout your repository so that all necessary files are available in the workflow.

   ```yaml
   - name: Checkout Repository
     uses: actions/checkout@v2
   ```

9. **Environment Isolation**: GitHub Actions workflows run in isolated environments. Ensure that the necessary dependencies and files are available in the runner's environment.

10. **Debugging Remote Runner**: If you suspect issues with the runner environment, you can SSH into the runner machine for debugging. GitHub provides instructions for this: [Debugging Actions](https://docs.github.com/en/actions/guides/troubleshooting#debugging-actions).

By following these steps and adding debug statements, you should be able to identify the root cause of the YAML unmarshaling error and resolve it in your GitHub Actions workflow.

**Comment by askb** [comment, marked as answer]
Using templates and reusable workflows would nest the job on multiple levels. This causes the secrets not to be passed to the referenced workflows correctly. However, the workaround I found was to set secrets through the environment vars and pass them to the referenced workflow works.
